service/round.go

```diff
@@ -191,108 +145,75 @@ func (r *round) persistExecution(tree *merkle.Tree, treeCache *cache.Writer, num
+	logger := logging.FromContext(ctx).WithFields(log.String("round", r.ID))
+	logger.With().Info("recovering execution", log.Field(zap.Time("end", end)))
+
+	started := time.Now()
 
-	if state.Members != nil && state.Statement != nil {
-		r.execution.Members = state.Members
-		r.execution.Statement = state.Statement
-	} else {
-		var err error
-		r.execution.Members, r.execution.Statement, err = r.calcMembersAndStatement()
+	if r.Execution.Members == nil || r.Execution.Statement == nil {
+		logger.Debug("calculating members and statement")
+		members, statement, err := r.calcMembersAndStatement()
 		if err != nil {
-			return err
+			return fmt.Errorf("failed to calculate members and statement")
 		}
+		r.Execution.Members, r.Execution.Statement = members, statement
 		if err := r.saveState(); err != nil {
 			return err
 		}
 	}
 
-	var err error
-	r.execution.NumLeaves, r.execution.NIP, err = prover.GenerateProofRecovery(
+	numLeaves, nip, err := prover.GenerateProofRecovery(
 		ctx,
 		r.datadir,
-		hash.GenLabelHashFunc(state.Statement),
-		hash.GenMerkleHashFunc(state.Statement),
+		hash.GenLabelHashFunc(r.Execution.Statement),
+		hash.GenMerkleHashFunc(r.Execution.Statement),
 		end,
-		state.SecurityParam,
-		state.NumLeaves,
-		state.ParkedNodes,
+		r.Execution.SecurityParam,
+		r.Execution.NumLeaves,
+		r.Execution.ParkedNodes,
 		r.persistExecution,
 	)
 	if err != nil {
 		return err
 	}
+	r.Execution.NumLeaves, r.Execution.NIP = numLeaves, nip
 	if err := r.saveState(); err != nil {
 		return err
 	}
 
-	close(r.executionEndedChan)
+	logger.With().Info("finished round recovered execution", log.Field(zap.Duration("duration", time.Since(started))))
 
 	return nil
 }
 
-func (r *round) proof(wait bool) (*PoetProof, error) {
-	if wait {
-		<-r.executionEndedChan
-	} else {
-		select {
-		case <-r.executionEndedChan:
-		default:
-			select {
-			case <-r.executionStartedChan:
-				return nil, errors.New("round is executing")
-			default:
-				select {
-				case <-r.openedChan:
-					return nil, errors.New("round is open")
-				default:
-					return nil, errors.New("round wasn't open")
-				}
-			}
-		}
-	}
-
-	return &PoetProof{
-		N:         uint(r.execution.NumLeaves),
-		Statement: r.execution.Statement,
-		Proof:     r.execution.NIP,
-	}, nil
-}
-
-func (r *round) state() (*roundState, error) {
+// loadState recovers persisted state from disk.
+func (r *round) loadState() error {
 	filename := filepath.Join(r.datadir, roundStateFileBaseName)
-	s := &roundState{}
-
-	if err := load(filename, s); err != nil {
-		return nil, err
+	state := roundState{}
+	if err := load(filename, &state); err != nil {
+		return err
 	}
-	if r.execution.SecurityParam != s.Execution.SecurityParam {
-		return nil, errors.New("SecurityParam config mismatch")
+	if r.Execution.SecurityParam != state.Execution.SecurityParam {
+		return errors.New("SecurityParam config mismatch")
 	}
-	r.stateCache = s
+	r.roundState = state
```

Comment:
Suggestion: the type `roundState` is only used for (de-)serialization of the state to/from disk but is not a field of the round.

Instead the round has `ExecutionStarted` and `Execution`  as private fields which are set here.

Reply:
I know. It was a conscious choice to put the _persistable_ state together in a struct. This way it's immediately clear what the _persistable_ state is. It's embedded anyway so there is no _visual_ cost of having to go through an extra level of nesting to work with the data.

Reply:
I can't follow your argument? What does the layout of the json file used for persistence have to do with the structure of the round object? Why does this have to be the same?

No part of the code besides `loadState()` and `saveState()` need to know what is persisted and how.

Reply:
The layout of xdr file has nothing to do with it. The argument is that **it is clear what data is persisted**.

Reply:
That's clear without having a subtype inside the round imo, also why does this need to be visible outside of the two methods I stated?

Reply:
It's less clear to me. Why is this such an issue?

Reply:
I don't see the benefit of having `roundstate` as an object. You say so it's clear what is persisted, but it comes at the drawback that anyone with access to round can modify its `ExecutionState` directly, which was not possible before.

`loadState()` and `saveState()` were very straight forward before and showed which data was persisted and how; this is now hidden away in `roundstate` with the additional drawback stated above. I don't see the benefit of this change.

Reply:
I see a benefit as explained.

Regarding:
> it comes at the drawback that anyone with access to round can modify its ExecutionState directly, which was not possible before.

It is not true. It has always been possible to modify the state from the outside as there is no concept of "private" state in Golang whatsoever. Everybody from `package service` can and always could modify the state of `round struct`. And as it is unexported, no one from outside of `package service` can hold a `round`. Please correct me if I'm wrong and why.

Reply:
Yes it is true. ExecutionState can be modified by anyone. It is public now and wasn't before. I am aware that before everyone inside the `package service` could modify it, but now everyone in any package can modify this value. And `round` doesn't need to be exported to be used outside of the package. I sent you an exact example.

There is no need to DRY up this code. I prefer the code how it was before and it didn't have the drawback of exposing the state publicly.

Reply:
changed.